This workbook implements Context Based Filtering for a Cats Recommendation System, working from the raw data all the way to model creation and initial results.

# Table of Contents

* [Load in Data and Segment Features from Context data](#segment)
* [Pre-process feature data](#pre-process)
    - [Make Needed Helper Functions and Imports](#pp_pipeline)
    - [Preprocess Data for model runs](#pp)
* [Run Content-Based-Filtering Modeling Iterations](#run_pipeline)
    - [Cosine similarity results](#cs)
    - [Overall Content Based Filtering results as of 1/18/2023](#ov)  
* [Conclusion and Next Steps](#conclusion)

# Load in Data and Segment Features from Context data<a id='segment'></a>

First, we load all of our adoptable cats.

In [ ]:
from google.colab import drive
import joblib #so I can save files out

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
cats_DF = pd.read_csv("/content/drive/MyDrive/MLE10PetMatch/Adoptable_cats_20221125.csv",header=0,index_col=0)
cats_DF.shape

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (42) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


(49600, 50)

In [ ]:
pd.set_option('display.max_columns', 500)
cats_DF.sample(3)

id organization_id  \
48711  58667767           GA423   
40050  58754877            ND32   
9683   58946253           WI207   

                                                     url type species     age  \
48711  https://www.petfinder.com/cat/rosemary-5866776...  Cat     Cat    Baby   
40050  https://www.petfinder.com/cat/luna-58754877/nd...  Cat     Cat    Baby   
9683   https://www.petfinder.com/cat/liberty-58946253...  Cat     Cat  Senior   

       gender    size   coat         tags      name  \
48711  Female   Small    NaN           []  Rosemary   
40050  Female   Small    NaN  ['Playful']      Luna   
9683   Female  Medium  Short           []   Liberty   

                                             description  \
48711                                                NaN   
40050  Hi. I&amp;#39;m Luna I am a little shy but ver...   
9683   Meet Liberty. She is a 7 year old cat looking ...   

      organization_animal_id  \
48711               51007780   
40050             FFR-A-3945   
9683                     NaN   

                                                  photos  \
48711  [{'small': 'https://dl5zpyw5k3jeb.cloudfront.n...   
40050  [{'small': 'https://dl5zpyw5k3jeb.cloudfront.n...   
9683   [{'small': 'https://dl5zpyw5k3jeb.cloudfront.n...   

       primary_photo_cropped videos     status         status_changed_at  \
48711                    NaN     []  adoptable  2022-10-26T19:26:19+0000   
40050                    NaN     []  adoptable  2022-11-04T13:21:28+0000   
9683                     NaN     []  adoptable  2022-11-23T16:15:16+0000   

                   published_at  distance        breeds.primary  \
48711  2022-10-26T19:26:19+0000       NaN   Domestic Short Hair   
40050  2022-11-04T13:21:27+0000       NaN  Domestic Medium Hair   
9683   2022-11-23T16:15:15+0000       NaN   Domestic Short Hair   

      breeds.secondary  breeds.mixed  breeds.unknown colors.primary  \
48711              NaN         False           False            NaN   
40050              NaN         False           False          Black   
9683               NaN         False           False            NaN   

      colors.secondary colors.tertiary  attributes.spayed_neutered  \
48711              NaN             NaN                       False   
40050              NaN             NaN                        True   
9683               NaN             NaN                        True   

       attributes.house_trained  attributes.declawed  \
48711                     False                False   
40050                     False                False   
9683                       True                False   

       attributes.special_needs  attributes.shots_current  \
48711                     False                     False   
40050                     False                      True   
9683                      False                      True   

      environment.children environment.dogs environment.cats  \
48711                  NaN              NaN              NaN   
40050                  NaN              NaN              NaN   
9683                   NaN              NaN             True   

                           contact.email   contact.phone  \
48711  adoption@dekalbanimalservices.com  (404) 294-2165   
40050                  ffrrinc@gmail.com             NaN   
9683          ocontoareahumane@gmail.com    920-835-1738   

        contact.address.address1 contact.address.address2  \
48711  3280 Chamblee Dunwoody Rd                      NaN   
40050                        NaN                      NaN   
9683          150 S. Katch Drive                      NaN   

      contact.address.city contact.address.state contact.address.postcode  \
48711             Chamblee                    GA                    30341   
40050             Bismarck                    ND                    58507   
9683                Oconto                    WI                    54153   

      contact.address.country  animal_id animal_type 

In [ ]:
cats_DF.columns

Index(['id', 'organization_id', 'url', 'type', 'species', 'age', 'gender',
       'size', 'coat', 'tags', 'name', 'description', 'organization_animal_id',
       'photos', 'primary_photo_cropped', 'videos', 'status',
       'status_changed_at', 'published_at', 'distance', 'breeds.primary',
       'breeds.secondary', 'breeds.mixed', 'breeds.unknown', 'colors.primary',
       'colors.secondary', 'colors.tertiary', 'attributes.spayed_neutered',
       'attributes.house_trained', 'attributes.declawed',
       'attributes.special_needs', 'attributes.shots_current',
       'environment.children', 'environment.dogs', 'environment.cats',
       'contact.email', 'contact.phone', 'contact.address.address1',
       'contact.address.address2', 'contact.address.city',
       'contact.address.state', 'contact.address.postcode',
       'contact.address.country', 'animal_id', 'animal_type',
       'organization_id.1', 'primary_photo_cropped.small',
       'primary_photo_cropped.medium', 'primary_photo

Drop animals with no pictures since they are key to our 'tinder-like' app experience.

In [ ]:
cats_DF = cats_DF.dropna(subset=['primary_photo_cropped.full'])# drop rows with 0 pictures
cats_DF.shape # matches na count via sweet viz for cats

(46805, 50)

Next we seperate the dataframe into features to model over and context data that can be shown to the user for any matches. 'ID' will be our shared key between the two tables.

Of note, the 'distance' field and 'primary_photo_cropped.full' field will be useful data for future model enhancements. For the models so far, we will simply use textual data and assume a 0 distance for all pets.

In [ ]:
contextCols = ['id','organization_id','url','type','tags','name','description','organization_animal_id',
              'photos','primary_photo_cropped','videos','status','status_changed_at','published_at',
              'distance','contact.email', 'contact.phone', 'contact.address.address1',
               'contact.address.address2', 'contact.address.city','contact.address.state', 
               'contact.address.postcode','contact.address.country', 'animal_id', 'animal_type',
               'organization_id.1', 'primary_photo_cropped.small','primary_photo_cropped.medium',
               'primary_photo_cropped.large','primary_photo_cropped.full']
featureCols = ['id','age','gender','size','coat','breeds.primary', 'breeds.secondary','breeds.mixed',
              'breeds.unknown','colors.primary','colors.secondary','colors.tertiary',
              'attributes.spayed_neutered','attributes.house_trained','attributes.declawed',
              'attributes.special_needs','attributes.shots_current','environment.children',
              'environment.dogs','environment.cats','type','contact.address.postcode'] # initial columns to keep for training purposes
cats_DF_features = cats_DF[featureCols]
cats_DF_context = cats_DF[contextCols]
cats_DF_features.shape

(46805, 22)

Let's sanity check our missing values now that we just have cats and remove any columns with too many missing values.

In [ ]:
valueCounts = cats_DF_features.set_index('type').isna().groupby(level=0).sum()/cats_DF_features.shape[0] # level=0 refers to our index, which we made 'type'


In [ ]:
pd.set_option('display.max_columns', 500)
valueCounts 

id  age  gender  size      coat  breeds.primary  breeds.secondary  \
type                                                                       
Cat   0.0  0.0     0.0   0.0  0.616857             0.0          0.899925   

      breeds.mixed  breeds.unknown  colors.primary  colors.secondary  \
type                                                                   
Cat            0.0             0.0        0.393548          0.746523   

      colors.tertiary  attributes.spayed_neutered  attributes.house_trained  \
type                                                                          
Cat          0.916035                         0.0                       0.0   

      attributes.declawed  attributes.special_needs  attributes.shots_current  \
type                                                                            
Cat                   0.0                       0.0                       0.0   

      environment.children  environment.dogs  environment.cats  \
type                                                             
Cat               0.737015          0.829954          0.588612   

      contact.address.postcode  
type                            
Cat                   0.000021

In [ ]:
valueCounts = cats_DF_context.set_index('type').isna().groupby(level=0).sum()/cats_DF_context.shape[0] # level=0 refers to our index, which we made 'type'


In [ ]:
pd.set_option('display.max_columns', 500)
valueCounts 

id  organization_id  url  tags      name  description  \
type                                                           
Cat   0.0              0.0  0.0   0.0  0.000021     0.262365   

      organization_animal_id  photos  primary_photo_cropped  videos  status  \
type                                                                          
Cat                 0.313022     0.0                    1.0     0.0     0.0   

      status_changed_at  published_at  distance  contact.email  contact.phone  \
type                                                                            
Cat                 0.0           0.0       1.0       0.051405       0.193804   

      contact.address.address1  contact.address.address2  \
type                                                       
Cat                   0.371499                  0.923534   

      contact.address.city  contact.address.state  contact.address.postcode  \
type                                                                          
Cat                    0.0                    0.0                  0.000021   

      contact.address.country  animal_id  animal_type  organization_id.1  \
type                                                                       
Cat                       0.0        0.0          0.0                0.0   

      primary_photo_cropped.small  primary_photo_cropped.medium  \
type                                                              
Cat                           0.0                           0.0   

      primary_photo_cropped.large  primary_photo_cropped.full  
type                                                           
Cat                           0.0                         0.0

After a quick NA check, we will have to remove 'coat','breeds.secondary','colors.secondary','colors.tertiary'. The column 'colors.primary' is also missing a lot of values but for sake of differing one cat from another it will be kept for now. Additionally, we will bring back in address postcode as an initial attempt to match nearby cats together. Lastly, 'environment.children','environment.dogs', and 'environment.cats' have a lot of missing values but users derive a lot of value from this information. Therefore, they will be kept as well.

In [ ]:
featureCols = ['id','age','gender','size','breeds.primary','breeds.mixed','breeds.unknown',
               'colors.primary','attributes.spayed_neutered','attributes.house_trained',
               'attributes.declawed','attributes.special_needs','attributes.shots_current',
               'contact.address.postcode','environment.children','environment.dogs','environment.cats']
cats_DF_features = cats_DF[featureCols]
cats_DF_context = cats_DF[contextCols]
cats_DF_features.shape

(46805, 17)

In [ ]:
cats_DF_features.dtypes

id                             int64
age                           object
gender                        object
size                          object
breeds.primary                object
breeds.mixed                    bool
breeds.unknown                  bool
colors.primary                object
attributes.spayed_neutered      bool
attributes.house_trained        bool
attributes.declawed             bool
attributes.special_needs        bool
attributes.shots_current        bool
contact.address.postcode      object
environment.children          object
environment.dogs              object
environment.cats              object
dtype: object

# Pre-process feature data<a id='pre-process'></a>

## Make Needed Helper Functions and Imports <a id='pp_pipeline'></a>

Make needed helper functions for modeling later on in workbook.

In [ ]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.metrics.pairwise import linear_kernel 
from sklearn.metrics.pairwise import laplacian_kernel
from sklearn.model_selection import train_test_split
import numpy as np
from sklearn.decomposition import TruncatedSVD

In [ ]:
def remove_columns_with_1_distinct(df):
    drop_col = [e for e in df.columns if df[e].nunique()==1]
    df_return = df.drop(drop_col,axis=1)
    return df_return


In [ ]:
def drop_duplicates(df):
    df_return = df.drop_duplicates()
    return df_return


In [ ]:
def linear_similarities(df_1,id_df):
    cs_simil = linear_kernel(df_1,df_1)
    results = {}
    ds = id_df # needs id column
    for idx, row in ds.iterrows():
       similar_indices = cs_simil[idx].argsort()[:-100:-1] 
       similar_items = [(cs_simil[idx][i], ds['id'][i]) for i in similar_indices] 
       results[row['id']] = similar_items[1:]
    return results

In [ ]:
def cosine_similarities(df_1,id_df):
    cs_simil = cosine_similarity(df_1,df_1)
    results = {}
    ds = id_df # needs id column
    for idx, row in ds.iterrows():
       similar_indices = cs_simil[idx].argsort()[:-100:-1] 
       similar_items = [(cs_simil[idx][i], ds['id'][i]) for i in similar_indices] 
       results[row['id']] = similar_items[1:]
    return results

In [ ]:
def laplacian_similarities(df_1,id_df):
    cs_simil = laplacian_kernel(df_1,df_1)
    results = {}
    ds = id_df # needs id column
    for idx, row in ds.iterrows():
       similar_indices = cs_simil[idx].argsort()[:-100:-1] 
       similar_items = [(cs_simil[idx][i], ds['id'][i]) for i in similar_indices] 
       results[row['id']] = similar_items[1:]
    return results

In [ ]:
def item(id,df):  
    ds = df
    colsGrab = ['id']
    return ds.loc[ds['id'] == id][colsGrab].values[0]# Just reads the results out of the dictionary.

def url(id,df):  
    ds = df
    colsGrab = ['url']
    return ds.loc[ds['id'] == id][colsGrab].values[0]# Just reads the results out of the dictionary.

def picture(id,df):  
    ds = df
    colsGrab = ['primary_photo_cropped.full']
    return ds.loc[ds['id'] == id][colsGrab].values[0]# Just reads the results out of the dictionary.

def recommend(item_id, num,df,reccs):
    print("Recommending " + str(num) + " cats similar to " + str(item(item_id,df)) + "... " 
          + picture(item_id,df) + " - " + url(item_id,df))   
    print("-------")    
    recs = reccs[item_id][:num]   
    for rec in recs: 
        print("Recommended: " + str(item(rec[1],df)) + " (score:" +      str(rec[0]) + ") " 
              + picture(rec[1],df) + " - " + url(rec[1],df))
    
def score(reccs, num):
    print("Finding average reccomendation score for top 5 reccomendations per example")
    results = []
    for key in reccs.keys():
        subRecs = reccs[key][:num]
        for r in subRecs:
            results.append(r[0])
    averageRecc = sum(results) / len(results)
    print("There are "+ str(len(results)) + 'results with a sum of' + str(sum(results)) + 'and and average of: ' 
          + str(averageRecc) )
    return averageRecc

## Preprocess Data for model runs <a id='pp'></a>

Now that essential methods are defined, lets handle the data.

In [ ]:
cats_DF_features.head(3) #sneak peak of what we have to work with initially

id    age  gender    size       breeds.primary  breeds.mixed  \
1   58980784   Baby    Male  Medium               Tuxedo         False   
13  58980778   Baby    Male  Medium  Domestic Short Hair         False   
14  58980506  Young  Female  Medium  Domestic Short Hair         False   

    breeds.unknown          colors.primary  attributes.spayed_neutered  \
1            False  Black & White / Tuxedo                        True   
13           False                   Black                        True   
14           False                  Torbie                        True   

    attributes.house_trained  attributes.declawed  attributes.special_needs  \
1                       True                False                     False   
13                      True                False                     False   
14                      True                False                     False   

    attributes.shots_current contact.address.postcode environment.children  \
1                       True                    37343                 True   
13                      True                    92057                 True   
14                      True                    50126                  NaN   

   environment.dogs environment.cats  
1               NaN             True  
13              NaN             True  
14              NaN             True

Besides the id value, which is our shared key, all other fields are categorical. We can use One-Hot encoding to transform them into something more efficient to run models over. 

Some preprocessing before One-Hot Encoding must occur to ensure everything goes as planned. First, we proactively drop duplicate rows. Second, we remove any features with only 1 distince value, since content-based filtering uses differences between objects and if everyone is the same there is no new information. Third, we replace NaNs with a special string so that One-Hot Encoding can work. Lastly, we fix the postcode to a string so that One-Hot Encoding works properly.

In [ ]:
# Preprocess data before encoding occurs for some troublesome fields
X = cats_DF_features
X = drop_duplicates(X) # remove duplicate rows
X = remove_columns_with_1_distinct(X) # remove any features with only 1 distinct value
X["contact.address.postcode"]= X["contact.address.postcode"].astype(str) # fix postcode to be a str rather than an int
# One-Hot Encoder requires all strings or all ints, so bools are not strings
X['breeds.mixed'] = X['breeds.mixed'].map({True: 'True', False: 'False'}) 
X['attributes.spayed_neutered'] = X['attributes.spayed_neutered'].map({True: 'True', False: 'False'}) 
X['attributes.house_trained'] = X['attributes.house_trained'].map({True: 'True', False: 'False'}) 
X['attributes.declawed'] = X['attributes.declawed'].map({True: 'True', False: 'False'}) 
X['attributes.special_needs'] = X['attributes.special_needs'].map({True: 'True', False: 'False'}) 
X['attributes.shots_current'] = X['attributes.shots_current'].map({True: 'True', False: 'False'}) 
X['environment.children'] = X['environment.children'].map({True: 'True', False: 'False'}) 
X['environment.dogs'] = X['environment.dogs'].map({True: 'True', False: 'False'}) 
X['environment.cats'] = X['environment.cats'].map({True: 'True', False: 'False'}) 
X = X.replace(np.nan,'Not Available') # replace nan's with their own special category, do this last once types all fixed!
X.dtypes

id                             int64
age                           object
gender                        object
size                          object
breeds.primary                object
breeds.mixed                  object
colors.primary                object
attributes.spayed_neutered    object
attributes.house_trained      object
attributes.declawed           object
attributes.special_needs      object
attributes.shots_current      object
contact.address.postcode      object
environment.children          object
environment.dogs              object
environment.cats              object
dtype: object

In [ ]:
X_transform = X

In [ ]:
X_transform.shape

(46710, 16)

In [ ]:
X_transform.head(3)

id    age  gender    size       breeds.primary breeds.mixed  \
1   58980784   Baby    Male  Medium               Tuxedo        False   
13  58980778   Baby    Male  Medium  Domestic Short Hair        False   
14  58980506  Young  Female  Medium  Domestic Short Hair        False   

            colors.primary attributes.spayed_neutered  \
1   Black & White / Tuxedo                       True   
13                   Black                       True   
14                  Torbie                       True   

   attributes.house_trained attributes.declawed attributes.special_needs  \
1                      True               False                    False   
13                     True               False                    False   
14                     True               False                    False   

   attributes.shots_current contact.address.postcode environment.children  \
1                      True                    37343                 True   
13                     True                    92057                 True   
14                     True                    50126        Not Available   

   environment.dogs environment.cats  
1     Not Available             True  
13    Not Available             True  
14    Not Available             True

In [ ]:
cats_DF_context.head(3)

id organization_id  \
1   58980784           TN589   
13  58980778          CA2825   
14  58980506            IA16   

                                                  url type  \
1   https://www.petfinder.com/cat/zorro-58980784/t...  Cat   
13  https://www.petfinder.com/cat/sammy-58980778/c...  Cat   
14  https://www.petfinder.com/cat/girly-58980506/i...  Cat   

                                                 tags    name  \
1   ['Friendly', 'Affectionate', 'Playful', 'Funny...  Zorro    
13  ['Friendly', 'Playful', 'Loves kisses', 'Athle...   Sammy   
14                ['Friendly', 'Gentle', 'Dignified']   Girly   

                                          description organization_animal_id  \
1   Zorro is very sweet and enjoys being in your l...                    NaN   
13  “Sammy” is a tiny black, male kitten. About 10...                    NaN   
14  Girly is a dainty lady! She enjoys getting pet...                      7   

                                               photos  primary_photo_cropped  \
1   [{'small': 'https://dl5zpyw5k3jeb.cloudfront.n...                    NaN   
13  [{'small': 'https://dl5zpyw5k3jeb.cloudfront.n...                    NaN   
14  [{'small': 'https://dl5zpyw5k3jeb.cloudfront.n...                    NaN   

                                               videos     status  \
1                                                  []  adoptable   
13                                                 []  adoptable   
14  [{'embed': '<iframe title="Video" frameborder=...  adoptable   

           status_changed_at              published_at  distance  \
1   2022-11-28T02:29:46+0000  2022-11-28T02:29:45+0000       NaN   
13  2022-11-28T02:29:10+0000  2022-11-28T02:29:09+0000       NaN   
14  2022-11-28T02:28:45+0000  2022-11-28T02:28:44+0000       NaN   

                  contact.email   contact.phone contact.address.address1  \
1           bulldog50@epbfi.com             NaN                      NaN   
13       info@sunriserescue.com             NaN                      NaN   
14  greenbelthumane@hotmail.com  (641) 648-2692            319 River St.   

   contact.address.address2 contact.address.city contact.address.state  \
1                       NaN               Hixson                    TN   
13                      NaN            Oceanside                    CA   
14                      NaN           Iowa Falls                    IA   

   contact.address.postcode contact.address.country  animal_id animal_type  \
1                     37343                      US   58980784         cat   
13                    92057                      US   58980778         cat   
14                    50126                      US   58980506         cat   

   organization_id.1                        primary_photo_cropped.small  \
1              tn589  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   
13            ca2825  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   
14              ia16  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   

                         primary_photo_cropped.medium  \
1   https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   
13  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   
14  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   

                          primary_photo_cropped.large  \
1   https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   
13  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   
14  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...   

                           primary_photo_cropped.full  
1   https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...  
13  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...  
14  https://dl5zpyw5k3jeb.cloudfront.net/photos/pe...

Since we are using unsupervised content-based filtering, the full data set will be used to create the model. No data splits will be used. 

The entries match! We need to pass to our models the numerical data to analyze similarity of products and the context data that goes along with it. As long as the indexes are the same, we can stitch them back together.

Since we know the indexs match, lets get rid of the id columns.

In [ ]:
X_transform = X_transform.reset_index(drop=True) # required so keys work properly
X_transform_woID = X_transform.drop(columns='id')
X_transform_woID.dtypes

age                           object
gender                        object
size                          object
breeds.primary                object
breeds.mixed                  object
colors.primary                object
attributes.spayed_neutered    object
attributes.house_trained      object
attributes.declawed           object
attributes.special_needs      object
attributes.shots_current      object
contact.address.postcode      object
environment.children          object
environment.dogs              object
environment.cats              object
dtype: object

Notice that indexes are the same and id columns are gone, so we can recover the IDs later! Now we can apply One-Hot Encoding!

In [ ]:
ohe = OneHotEncoder().fit(X_transform_woID) # One Hot Encoding WAAAY better, fit on whole X
X_train_transform = ohe.transform(X_transform_woID) # don't need to add id columns because same columns preserved

# Run Content-Based-Filtering Modeling Iterations <a id='run_pipeline'></a>

Content-Based Filtering is a method of comparing products against each other when you don't have user rankings. This can be a simple way to create models before user ranking data is available and can often do well in recommending similar products. In our case, products are cats. Cosine Similarity will be used moving forward.

## Cosine similarity results <a id='cs'></a>

In [ ]:
Cosine_Model =cosine_similarities(X_train_transform,X_transform) #run similarities with cosine similarity


In [ ]:
joblib.dump(Cosine_Model, '/content/drive/MyDrive/MLE10PetMatch/models/cosine_similarity_model_catsv2.pkl')

['/content/drive/MyDrive/MLE10PetMatch/models/cosine_similarity_model_catsv2.pkl']

In [ ]:
Cosine_Model = joblib.load('/content/drive/MyDrive/MLE10PetMatch/models/cosine_similarity_model_catsv2.pkl')

In [ ]:
pd.options.display.max_colwidth = 100
recommend(item_id=58706766, num=5,df=cats_DF_context,reccs=Cosine_Model)

['Recommending 5 cats similar to [58706766]... https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/58706766/2/?bust=1667159800 - https://www.petfinder.com/cat/pjs-58706766/tx/austin/new-hope-animal-rescue-nfp-tx2339/?referrer_id=c2f7479c-c7e8-422b-bfb4-7c0b8aed0e55']
-------
['Recommended: [58926364] (score:0.9333333333333331) https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/58926364/1/?bust=1669040782 - https://www.petfinder.com/cat/squidlet-58926364/va/alexandria/tails-high-inc-va540/?referrer_id=c2f7479c-c7e8-422b-bfb4-7c0b8aed0e55']
['Recommended: [58706847] (score:0.9333333333333331) https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/58706847/2/?bust=1667160445 - https://www.petfinder.com/cat/sleepy-spice-58706847/tx/austin/new-hope-animal-rescue-nfp-tx2339/?referrer_id=c2f7479c-c7e8-422b-bfb4-7c0b8aed0e55']
['Recommended: [58811866] (score:0.9333333333333331) https://dl5zpyw5k3jeb.cloudfront.net/photos/pets/58811866/1/?bust=1668884795 - https://www.petfinder.com/cat/bria-58811866/az/q

The above is score for one item only so now let's get an idea of how well this does for the entire training set.

In [ ]:
# Gather average score of top 5 recommendations for training set, with a max score of 1!
cosineScore = score(reccs=Cosine_Model, num=5)
cosineScore

Finding average reccomendation score for top 5 reccomendations per example
There are 233550results with a sum of215295.19999944614and and average of: 0.9218377221128072


0.9218377221128072

The overall score for the whole training set for Cosine Similarity is .922

## Overall Content Based Filtering results as of 1/18/2023 <a id='ov'></a>

In [ ]:
from tabulate import tabulate
table = [['Model Name', 'Score'],
         ['Cosine Similarity',cosineScore]]
print(tabulate(table,headers='firstrow',tablefmt='fancy_grid'))

╒═══════════════════╤══════════╕
│ Model Name        │    Score │
╞═══════════════════╪══════════╡
│ Cosine Similarity │ 0.921838 │
╘═══════════════════╧══════════╛


# Conclusion and Next Steps <a id='conclusion'></a>

**Conclusion of ML Modeling as of 1/18/23:**


*   Cosine Similarity with all examples improves the performance by .004
*   Will now work with UI and full cat set



**Conclusion of ML Modeling as of 1/2/23**: 
- All three content-based filtering models perform well
- Cosine Similarity appears to be the most sensitive to differences and has a very useful scale of 0-1.
- Can hook up content-based filtering models to PetMatch UI as-is and it should return good results based on overall similarity measures measured so far.
- User Rankings Data generated require more formating than initially expected but our application tracks all the key required fields for now.
- Collaborative Filtering is harder to implement than initially expected, but we have initial data to give it a try.

**Conclusion of ML Baseline as of 12/6/22**: 
- Average top 5 recommendation per cat in the training set is 10.96. The highest available score is a 12.  
- The result above uses a simple content-based filtering recommendation model without using user perferences, since they are currently not available. Instead it compares items against each other, aka you liked this ketchup so here are 10 other similar types of ketchup. 
- Due to the method used to create the simple content-based filtering model, dev and test set can not be used so to get an initial idea of the results the training set was used. 
- The cats data version 0.5 features need more ways to dileanate one cat from another but based on include visual scans and the average reccomendation score, the simple cat CBF model generally excels at giving you similar cats to what you stated you wanted.
- In instances where there is more ambiguity (aka a chosen cat with less defined details), it will still find cats very similar to it but sometimes it can also throw in very similar cats who are a different breed. This might not be a bad thing.

**Next Steps**:

- Get more user rankings!
- Incorporate distance more effectively
- Can we use description field for cats at all? 
- Collaborative Filtering for item and user-based
  - Use surprise library possibly
  - Add timestamp to rankings so we can be time-sensitive in terms of reccomendations